src/platform/unix/mod.rs

```diff
@@ -597,34 +598,26 @@ impl OsIpcOneShotServer {
     pub fn new() -> Result<(OsIpcOneShotServer, String),UnixError> {
         unsafe {
             let fd = libc::socket(libc::AF_UNIX, SOCK_SEQPACKET, 0);
-            let mut path: Vec<u8>;
-            let dir = TempDir::new().unwrap();
-            let dir_path = dir.path().join("rust-ipc-socket");
+            let _temp_dir =  Builder::new()
+                .tempdir()
+                .unwrap();
             
-            loop {
-
-                let path_string = CString::new(dir_path.to_str().unwrap()).unwrap();
-                path = path_string.as_bytes_with_nul().iter().cloned().collect();
-              
-              
-                let (sockaddr, len) = new_sockaddr_un(path.as_ptr() as *const c_char);
-                if libc::bind(fd, &sockaddr as *const _ as *const sockaddr, len as socklen_t) == 0 {
-                    break
-                }
+            let socket_path = _temp_dir.path().join("socket");
+            let path_string = CString::new(socket_path.to_str().unwrap()).unwrap();
 
-                let errno = UnixError::last();
-                if errno.0 != libc::EINVAL {
-                    return Err(errno)
-                }
+            let (sockaddr, len) = new_sockaddr_un(path_string.as_ptr() as *const c_char);
+            if libc::bind(fd, &sockaddr as *const _ as *const sockaddr, len as socklen_t) != 0 {
+                return Err(UnixError::last());
             }
 
             if libc::listen(fd, 10) != 0 {
                 return Err(UnixError::last())
             }
 
             Ok((OsIpcOneShotServer {
-                fd: fd, _dir: dir,
-            }, String::from_utf8(CStr::from_ptr(path.as_ptr() as
+                fd: fd,
+                _temp_dir: _temp_dir,
+            }, String::from_utf8(CStr::from_ptr(path_string.as_ptr() as
                                                 *const c_char).to_bytes().to_owned()).unwrap()))
```

Comment:
This complicated conversion is not necessary any more: you actually already have the final rust string after the `socket_path.to_str().unwrap()` (before converting it to `CString`) -- you just need to add a `.to_owned()`, and save it in another temporary variable. (Or even chain this conversion right in the assignment of `socket_path`.)